enhancements/machine-config/onprem-namespaces.md

```diff
@@ -0,0 +1,134 @@
+---
+title: namespaces-for-on-prem-networking-services
+authors:
+  - "@cybertron"
+reviewers:
+  - "@smarterclayton"
+  - "@celebdor"
+  - "@mandre"
+  - "@rgolangh"
+  - "@patrickdillon"
+  - "@kikisdeliveryservice"
+approvers:
+  - Someone from MCO
+  - Someone from architecture
+creation-date: 2020-07-24
+last-updated: 2020-07-27
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Namespaces for On-Prem Networking Services
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+When support for on-premise deployments on baremetal, OpenStack, OVirt, and
+VSphere was added, some additional networking services were needed to provide
+functionality that would come from cloud services on other platforms. These
+services were added in namespaces called openshift-kni-infra,
+openshift-openstack-infra, etc. There are two problems with this: 1) The
+namespaces are created in all deployments, including on platforms other
+than the one that uses the namespace and 2) The openshift-x-infra name scheme
+does not match the existing pattern for OpenShift namespaces.
+
+## Motivation
+
+The existence of namespaces not related to a given deployment is confusing to
+users. Additionally, the inconsistent naming of the namespaces is a further
+possible point of confusion.
+
+### Goals
+
+- Only create namespaces for networking services on deployments that need them.
+- Use names that are consistent with the rest of OpenShift.
+
+### Non-Goals
+
+## Proposal
+
+Instead of creating the namespaces at install time as is currently done, add
+them to the list of
+[resources to be synced](https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/sync.go)
+in the machine-config-operator. This will allow us to add logic that considers
+the platform before creating them.
+
+For the naming, we propose `openshift-hosted-network-services` for all
+platforms that use these components. There is no need for a platform-specific
+name since only one platform can exist at a time anyway.
+
+### User Stories [optional]
+
+### Implementation Details/Notes/Constraints [optional]
+
+The services in question provide things like internal DNS resolution between
+nodes and loadbalancing for API and ingress traffic. In cloud platforms these
+would normally be provided by a service in the cloud, but for baremetal and
+our other on-premise platforms there are no such services available.
+
+### Risks and Mitigations
+
+The only significant risk should be the migration from the existing names
+to the new ones. We are already creating namespaces for these services.
```

Comment:
> We are already creating namespaces for these services.
This proposal only changes how we do that and what those namespaces are
called.

I do not think it is that simple right? we need to migrate all our existing cluster from previous namespaces to this namespace as part of the upgrade.

Reply:
Maybe I'm wrong, but since these are static pods I thought that on upgrade MCO would write the new manifests and kubelet would kill the old version and start the new one. AFAIK we don't depend on anything in the namespace, it's just an accounting requirement for resource tracking.